File: DesignPatterns/IteratorPattern/NotificationClient.java
class Notification {

  private String notification;

  public Notification(String notification) {
    this.notification = notification;
  }

  public String getNotification() {
    return this.notification;
  }
}

interface Iterator{
  public boolean hasNext();
  public Object next();
}

class NotificationIterator implements Iterator {

  private Notification[] notificationArr;
  private int pos;

  public NotificationIterator(Notification[] notificationArr){
    this.notificationArr = notificationArr;
    this.pos = 0;
  }

  public Object next() {
    Notification notification = notificationArr[pos++];
    return notification;
  }

  public boolean hasNext(){
    if(pos >= notificationArr.length || notificationArr[pos] == null)
      return false;
    return true;
  }
}

interface Collection {
  public Iterator createIterator();
}

class NotificationCollection implements Collection {
  private static final int MAX_ITEMS = 6;
  private int numberOfItems = 0;
  private Notification[] notificationArr;

  public NotificationCollection(){
    notificationArr = new Notification[MAX_ITEMS];
  }

  public void addItem(String str){
    Notification notification = new Notification(str);
    if(numberOfItems >= MAX_ITEMS){
      System.out.println("Full");
    } else {
      notificationArr[numberOfItems] = notification;
      numberOfItems++;
    }
  }

  public Iterator createIterator(){
    return new NotificationIterator(notificationArr);
  }
}


class NotificationBar {

  private NotificationCollection notifications;

  public NotificationBar(NotificationCollection notifications){
    this.notifications = notifications;
  }

  public void printNotifications(){
    Iterator it = notifications.createIterator();
    System.out.println("Notification Bar ....");
    while(it.hasNext()){
      Notification nt = (Notification)it.next();
      System.out.println(nt.getNotification());
    }
  }
}


public class NotificationClient {
  public static void main(String[] args) {
    NotificationCollection ntc = new NotificationCollection();
    ntc.addItem("Notification-1");
    ntc.addItem("Notification-2");
    ntc.addItem("Notification-3");
    NotificationBar ntb = new NotificationBar(ntc);
    ntb.printNotifications();
  }
}

File: DesignPatterns/Behavioral/Iterator/Ruby/notification_client.rb
class Notification
  attr_reader :notification

  def initialize(notification)
    @notification = notification
  end
end

class Iterator
  attr_reader :pos

  def initialize
    @pos = 0
  end
end

class NotificationIterator < Iterator
  attr_reader :notification_arr

  def initialize(notification_arr)
    super()
    @notification_arr = notification_arr.clone
  end

  def next
    value = @notification_arr[@pos]
    @pos += 1
    value
  end

  def has_next?
    @pos >= @notification_arr.length || @notification_arr[@pos].nil? ? false : true
  end
end

class Collection
  attr_reader :items, :total_items

  @@max_items = 6

  def initialize
    @items = []
    @total_items = 0
  end
end

class NotificationCollection < Collection
  def add_item(item)
    if @total_items > @@max_items
      puts 'Full'
    else
      notification = Notification.new(item)
      @items << notification
      @total_items += 1
    end
  end

  def create_iterator
    NotificationIterator.new(@items)
  end
end

notification_collection = NotificationCollection.new
notification_collection.add_item("Notification-1")
notification_collection.add_item("Notification-2")
notification_collection.add_item("Notification-3")

notification_iterator = notification_collection.create_iterator
while notification_iterator.has_next? do
  puts notification_iterator.next.notification
end